partiql-eval/src/eval/expr/path.rs

```diff
@@ -96,13 +96,17 @@ impl EvalExpr for EvalDynamicLookup {
 
 /// Represents a variable reference in a (sub)query, e.g. `a` in `SELECT b as a FROM`.
 #[derive(Debug)]
-pub(crate) struct EvalVarRef {
-    pub(crate) name: BindingsName,
+pub(crate) enum EvalVarRef {
+    Local(BindingsName),
+    Global(BindingsName),
 }
 
 impl EvalExpr for EvalVarRef {
     fn evaluate<'a>(&'a self, bindings: &'a Tuple, ctx: &'a dyn EvalContext) -> Cow<'a, Value> {
-        let value = Bindings::get(bindings, &self.name).or_else(|| ctx.bindings().get(&self.name));
+        let value = match self {
+            EvalVarRef::Global(name) => ctx.bindings().get(name),
+            EvalVarRef::Local(name) => Bindings::get(bindings, name),
+        };
```

Comment:
As the variable ref type is known at compile-time, it would be better to not make every lookup execute a conditional.

Perhaps something like:
```rust
/// Represents a local variable reference in a (sub)query, e.g. `b` in `SELECT t.b as a FROM T as t`.
#[derive(Debug, Clone)]
pub(crate) enum EvalVarRef {
    Local(BindingsName),
    Global(BindingsName),
}

impl BindEvalExpr for EvalVarRef {
    fn bind<const STRICT: bool>(
        &self,
        _: Vec<Box<dyn EvalExpr>>,
    ) -> Result<Box<dyn EvalExpr>, BindError> {
        Ok(match self {
            EvalVarRef::Global(name) => Box::new(EvalGlobalVarRef { name: name.clone() }),
            EvalVarRef::Local(name) => Box::new(EvalLocalVarRef { name: name.clone() }),
        })
    }
}

#[inline]
fn borrow_or_missing<'a>(value: Option<&'a Value>) -> Cow<'a, Value> {
    value.map_or_else(|| Cow::Owned(Value::Missing), Cow::Borrowed)
}

/// Represents a local variable reference in a (sub)query, e.g. `b` in `SELECT t.b as a FROM T as t`.
#[derive(Debug)]
pub(crate) struct EvalLocalVarRef {
    pub(crate) name: BindingsName,
}

impl EvalExpr for EvalLocalVarRef {
    fn evaluate<'a>(&'a self, bindings: &'a Tuple, _: &'a dyn EvalContext) -> Cow<'a, Value> {
        borrow_or_missing(Bindings::get(bindings, &self.name))
    }
}

/// Represents a global variable reference in a (sub)query, e.g. `T` in `SELECT t.b as a FROM T as t`.
#[derive(Debug)]
pub(crate) struct EvalGlobalVarRef {
    pub(crate) name: BindingsName,
}

impl EvalExpr for EvalGlobalVarRef {
    fn evaluate<'a>(&'a self, _: &'a Tuple, ctx: &'a dyn EvalContext) -> Cow<'a, Value> {
        borrow_or_missing(ctx.bindings().get(&self.name))
    }
}
```

and planning might look like:

```rust
            ValueExpr::VarRef(name, var_ref_type) => (
                "var ref",
                match var_ref_type {
                    VarRefType::Global => EvalVarRef::Global(name.clone()),
                    VarRefType::Local => EvalVarRef::Local(name.clone()),
                }
                .bind::<{ STRICT }>(vec![]),
            ),
```

Reply:
Thanks for the suggestion. Updated to your suggestion in https://github.com/partiql/partiql-lang-rust/pull/416/commits/1748f5bd668a2e0096f29a72a24698973b34a11a.